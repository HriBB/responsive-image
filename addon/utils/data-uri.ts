export default function dataUri(
  data: string,
  type: string,
  base64 = false
): string {
  return `data:${type};base64,${base64 ? data : btoa(data)}`;
}
