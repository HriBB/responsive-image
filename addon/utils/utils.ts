export function normalizeSrc(src: string): string {
  return src[0] === '/' ? src.slice(1) : src;
}
