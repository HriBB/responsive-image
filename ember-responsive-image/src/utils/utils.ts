import Application, { getOwner } from '@ember/application';
import { assert } from '@ember/debug';
import { Meta } from '../types';

export function normalizeSrc(src: string): string {
  return src[0] === '/' ? src.slice(1) : src;
}

const SCRIPT_ID = 'ember_responsive_image_meta';

export function extractMeta(context: unknown): Meta {
  let script = document.getElementById(SCRIPT_ID);
  if (!script) {
    // in case the app runs in shadow DOM, we need to query the script from the shadow root
    const rootElement = (getOwner(context) as Application).rootElement;
    if (rootElement instanceof HTMLElement) {
      script = (rootElement.getRootNode() as HTMLElement).querySelector(
        `[id="${SCRIPT_ID}"]`
      );
    }
  }
  assert(
    'No script tag found containing meta data for ember-responsive-image',
    script && script.textContent // Cannot use optional chaining due to https://github.com/ember-cli/babel-plugin-debug-macros/issues/89
  );
  return JSON.parse(script.textContent);
}
