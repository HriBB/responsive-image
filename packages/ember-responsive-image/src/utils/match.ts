import type { ImageOutputResult, ImageType } from '../types.ts';

export function findMatchingImage(
  images: ImageOutputResult[],
  width: number,
  type: ImageType
): ImageOutputResult | undefined {
  const matchingImageWidth = images
    .map((i) => i.width)
    .reduce((prevValue: number, w: number) => {
      if (w >= width && prevValue >= width) {
        return w >= prevValue ? prevValue : w;
      } else {
        return w >= prevValue ? w : prevValue;
      }
    }, 0);

  return images.find(
    (image) => image.width === matchingImageWidth && image.format === type
  );
}
