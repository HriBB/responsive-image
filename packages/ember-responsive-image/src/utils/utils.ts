import Application, { getOwner } from '@ember/application';
import { assert } from '@ember/debug';
import { ImageOutputResult, ImageType, Meta } from '../types';

export function normalizeSrc(src: string): string {
  return src[0] === '/' ? src.slice(1) : src;
}

const SCRIPT_ID = 'ember_responsive_image_meta';

export function extractMeta(context: unknown): Meta {
  let script = document.getElementById(SCRIPT_ID);
  if (!script) {
    // in case the app runs in shadow DOM, we need to query the script from the shadow root
    const rootElement = (getOwner(context) as Application).rootElement;
    if (rootElement instanceof HTMLElement) {
      script = (rootElement.getRootNode() as HTMLElement).querySelector(
        `[id="${SCRIPT_ID}"]`
      );
    }
  }
  assert(
    'No script tag found containing meta data for ember-responsive-image',
    script && script.textContent // Cannot use optional chaining due to https://github.com/ember-cli/babel-plugin-debug-macros/issues/89
  );
  return JSON.parse(script.textContent);
}

export function findMatchingImage(
  images: ImageOutputResult[],
  width: number,
  type: ImageType
): ImageOutputResult | undefined {
  const matchingImageWidth = images
    .map((i) => i.width)
    .reduce((prevValue: number, w: number) => {
      if (w >= width && prevValue >= width) {
        return w >= prevValue ? prevValue : w;
      } else {
        return w >= prevValue ? w : prevValue;
      }
    }, 0);

  return images.find(
    (image) => image.width === matchingImageWidth && image.format === type
  );
}
