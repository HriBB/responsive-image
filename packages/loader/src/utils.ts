import sharp, { Metadata } from 'sharp';
import { LoaderContext } from 'webpack';
import {
  ImageLoaderChainedResult,
  LoaderOptions,
  LqipLoaderOptions,
} from './types';
import baseN from 'base-n';

const b64 = baseN.create();

const defaultImageConfig: LoaderOptions = {
  quality: 80,
  widths: [2048, 1536, 1080, 750, 640],
  formats: ['original', 'webp'],
  name: '[name]-[width]w-[hash].[ext]',
  outputPath: 'assets/images',
};

export function parseQuery(query: string): Record<string, unknown> {
  const params = new URLSearchParams(query);
  return Object.fromEntries(
    [...params.entries()].map(([key, value]) => {
      switch (key) {
        case 'widths':
          return [key, value.split(',').map((v) => parseInt(v, 10))];
        case 'formats':
          return [key, value.split(',')];
        case 'quality':
          return [key, parseInt(value, 10)];
        case 'lqip': {
          const parsedValue =
            value.charAt(0) === '{' ? JSON.parse(value) : value;

          return [
            key,
            typeof parsedValue === 'string'
              ? { type: parsedValue }
              : parsedValue,
          ];
        }

        default:
          return [key, value];
      }
    })
  );
}

export function getOptions(
  context: LoaderContext<Partial<LoaderOptions>>
): LoaderOptions {
  const parsedResourceQuery = parseQuery(context.resourceQuery);

  // Combines defaults, webpack options and query options,
  return {
    ...defaultImageConfig,
    ...context.getOptions(),
    ...parsedResourceQuery,
  };
}

export function normalizeInput(
  input: Buffer | ImageLoaderChainedResult
): ImageLoaderChainedResult {
  if (Buffer.isBuffer(input)) {
    return {
      images: [],
      sharp: sharp(input),
      imports: [],
    };
  }

  return input;
}

const generatedClassNames = new Map<string, string>();

export function generateLqipClassName(resource: string): string {
  if (generatedClassNames.has(resource)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return generatedClassNames.get(resource)!;
  } else {
    const className = `eri-dyn-${b64.encode(generatedClassNames.size)}`;
    generatedClassNames.set(resource, className);

    return className;
  }
}

export function getAspectRatio(meta: Metadata): number | undefined {
  if (meta.height && meta.width && meta.height > 0) {
    return Math.round((meta.width / meta.height) * 100) / 100;
  }
}

export function onlyUnique<T>(value: T, index: number, self: T[]): boolean {
  return self.indexOf(value) === index;
}

// export function normalizeInput(
//   image: Buffer,
//   meta?: ImageLoaderChainedResult
// ): ImageLoaderChainedResult {
//   if (meta) {
//     return meta;
//   }

//   return {
//     images: [],
//     data: sharp(image),
//   };
// }
