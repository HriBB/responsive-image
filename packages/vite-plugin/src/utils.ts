import baseN from 'base-n';
import sharp, { type Metadata } from 'sharp';
import type { ImageLoaderChainedResult, Options, ViteOptions } from './types';
import type { ImageConfig } from 'imagetools-core';
import { PluginContext } from 'rollup';
import { ResolvedConfig } from 'vite';

const b64 = baseN.create();

const defaultImageConfig: Options = {
  quality: 80,
  w: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
  // TODO: remove this, needs fixing tests
  allowUpscale: 'true',
  format: ['original', 'webp'],
  name: '[name]-[width]w.[ext]',
};

const webpackLoaderKeys: string[] = ['lqip', 'name'] satisfies Array<
  keyof ViteOptions
>;
const queryArraySeparator = ';';

export const META_KEY = 'responsive-image';

export function parseURL(id: string) {
  return new URL(id, 'file://');
}

export function parseQuery(query: string | URLSearchParams): Partial<Options> {
  const params =
    query instanceof URLSearchParams ? query : new URLSearchParams(query);
  return Object.fromEntries(
    [...params.entries()].map(([key, value]) => {
      switch (key) {
        case 'w':
          return [
            key,
            value.split(queryArraySeparator).map((v) => parseInt(v, 10)),
          ];
        case 'format':
          return [key, value.split(queryArraySeparator)];
        case 'quality':
          return [key, parseInt(value, 10)];
        case 'lqip': {
          const parsedValue =
            value.charAt(0) === '{' ? JSON.parse(value) : value;

          return [
            key,
            typeof parsedValue === 'string'
              ? { type: parsedValue }
              : parsedValue,
          ];
        }

        default:
          return [key, value];
      }
    }),
  );
}

export async function getImagetoolsConfigs(
  options: Options,
): Promise<ImageConfig[]> {
  const imagetools = await import('imagetools-core');

  const entries = Object.entries(options)
    .filter(([key]) => !webpackLoaderKeys.includes(key))
    .map(([key, value]) => {
      // imagetools expects this type
      const stringarrayifiedValue = Array.isArray(value)
        ? value.map((v) => String(v))
        : [String(value)];

      return [key, stringarrayifiedValue] satisfies [string, string[]];
    });

  return imagetools.resolveConfigs(entries, imagetools.builtinOutputFormats);
}

export function getOptions(
  id: string | URL,
  loaderOptions: Partial<Options>,
): Options {
  const url = id instanceof URL ? id : parseURL(id);
  const parsedResourceQuery = parseQuery(url.searchParams);

  // Combines defaults, webpack options and query options,
  return {
    ...defaultImageConfig,
    ...loaderOptions,
    ...parsedResourceQuery,
  } as Options;
}

export function normalizeInput(
  input: string | ImageLoaderChainedResult,
): ImageLoaderChainedResult {
  if (typeof input === 'string') {
    const url = parseURL(input);
    const pathname = decodeURIComponent(url.pathname);

    return {
      images: [],
      sharp: sharp(pathname),
      imports: [],
    };
  }

  return input;
}

export function getInput(
  context: PluginContext,
  id: string,
): ImageLoaderChainedResult | undefined {
  const info = context.getModuleInfo(id);
  return info?.meta[META_KEY];
}

const generatedClassNames = new Map<string, string>();

export function generateLqipClassName(resource: string): string {
  if (generatedClassNames.has(resource)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return generatedClassNames.get(resource)!;
  } else {
    const className = `eri-dyn-${b64.encode(generatedClassNames.size)}`;
    generatedClassNames.set(resource, className);

    return className;
  }
}

export function getAspectRatio(meta: Metadata): number | undefined {
  if (meta.height && meta.width && meta.height > 0) {
    return Math.round((meta.width / meta.height) * 100) / 100;
  }
}

export function onlyUnique<T>(value: T, index: number, self: T[]): boolean {
  return self.indexOf(value) === index;
}

export function dataUri(
  data: string | Buffer,
  type: string,
  base64 = false,
): string {
  return `data:${type};base64,${
    base64 ? data : Buffer.from(data).toString('base64')
  }`;
}

export function blurrySvg(src: string, width: number, height: number): string {
  return `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 ${width} ${height}">
<filter id="b" color-interpolation-filters="sRGB"><feGaussianBlur stdDeviation=".5"></feGaussianBlur><feComponentTransfer><feFuncA type="discrete" tableValues="1 1"></feFuncA></feComponentTransfer></filter>
<image filter="url(#b)" preserveAspectRatio="none" height="100%" width="100%" xlink:href="${src}"></image>
</svg>`;
}

export function getViteBasePath(viteConfig: ResolvedConfig): string {
  return `${
    viteConfig.base.endsWith('/')
      ? viteConfig.base.slice(0, -1)
      : viteConfig.base
  }/@responsive-image/vite-plugin/`;
}
