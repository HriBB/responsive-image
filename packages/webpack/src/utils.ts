import baseN from 'base-n';
import sharp, { Metadata } from 'sharp';
import { LoaderContext } from 'webpack';
import { ImageLoaderChainedResult, LoaderOptions } from './types';

const b64 = baseN.create();

const defaultImageConfig: LoaderOptions = {
  quality: 80,
  widths: [2048, 1536, 1080, 750, 640],
  formats: ['original', 'webp'],
  name: '[name]-[width]w-[hash].[ext]',
  outputPath: 'assets/images',
};

export function parseQuery(query: string): Record<string, unknown> {
  const params = new URLSearchParams(query);
  return Object.fromEntries(
    [...params.entries()].map(([key, value]) => {
      switch (key) {
        case 'widths':
          return [key, value.split(',').map((v) => parseInt(v, 10))];
        case 'formats':
          return [key, value.split(',')];
        case 'quality':
          return [key, parseInt(value, 10)];
        case 'lqip': {
          const parsedValue =
            value.charAt(0) === '{' ? JSON.parse(value) : value;

          return [
            key,
            typeof parsedValue === 'string'
              ? { type: parsedValue }
              : parsedValue,
          ];
        }

        default:
          return [key, value];
      }
    })
  );
}

export function getOptions(
  context: LoaderContext<Partial<LoaderOptions>>
): LoaderOptions {
  const parsedResourceQuery = parseQuery(context.resourceQuery);

  // Combines defaults, webpack options and query options,
  return {
    ...defaultImageConfig,
    ...context.getOptions(),
    ...parsedResourceQuery,
  };
}

export function normalizeInput(
  input: Buffer | ImageLoaderChainedResult
): ImageLoaderChainedResult {
  if (Buffer.isBuffer(input)) {
    return {
      images: [],
      sharp: sharp(input),
      imports: [],
    };
  }

  return input;
}

const generatedClassNames = new Map<string, string>();

export function generateLqipClassName(resource: string): string {
  if (generatedClassNames.has(resource)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return generatedClassNames.get(resource)!;
  } else {
    const className = `eri-dyn-${b64.encode(generatedClassNames.size)}`;
    generatedClassNames.set(resource, className);

    return className;
  }
}

export function getAspectRatio(meta: Metadata): number | undefined {
  if (meta.height && meta.width && meta.height > 0) {
    return Math.round((meta.width / meta.height) * 100) / 100;
  }
}

export function onlyUnique<T>(value: T, index: number, self: T[]): boolean {
  return self.indexOf(value) === index;
}

export function dataUri(
  data: string | Buffer,
  type: string,
  base64 = false
): string {
  return `data:${type};base64,${
    base64 ? data : Buffer.from(data).toString('base64')
  }`;
}

export function blurrySvg(src: string, width: number, height: number): string {
  return `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 ${width} ${height}">
<filter id="b" color-interpolation-filters="sRGB"><feGaussianBlur stdDeviation=".5"></feGaussianBlur><feComponentTransfer><feFuncA type="discrete" tableValues="1 1"></feFuncA></feComponentTransfer></filter>
<image filter="url(#b)" preserveAspectRatio="none" height="100%" width="100%" xlink:href="${src}"></image>
</svg>`;
}
