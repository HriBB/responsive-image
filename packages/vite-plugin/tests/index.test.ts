import { describe, expect, test } from 'vitest';
import { configureToMatchImageSnapshot } from 'jest-image-snapshot';
import { compile } from './utils';

const customConfig = { threshold: 0.1 };
const toMatchImageSnapshot = configureToMatchImageSnapshot({
  customDiffConfig: customConfig,
});

expect.extend({ toMatchImageSnapshot });

describe('filter', () => {
  test('it operates on included assets', async () => {
    const { source } = await compile('image.jpg', {
      include: '**/*.jpg',
    });

    expect(source).toMatchSnapshot();
  });

  test('it skips non-matches', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.png',
    });

    expect(source).not.toContain('imageUrlFor');
    expect(assets).toHaveLength(0);
  });

  test('it skips excluded assets', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.jpg',
      exclude: '**/image.jpg',
    });

    expect(source).not.toContain('imageUrlFor');
    expect(assets).toHaveLength(0);
  });
});

test('it produces expected output', async () => {
  const { source, assets } = await compile('image.jpg', {
    format: ['png', 'webp'],
    include: '**/*.jpg',
  });

  expect(source).toMatchSnapshot();

  const sortedAssets = assets.toSorted((a, b) =>
    a.fileName.localeCompare(b.fileName),
  );

  expect(sortedAssets.map((a) => a.fileName)).toEqual(
    [
      'images/image-640w.png',
      'images/image-640w.webp',
      'images/image-750w.png',
      'images/image-750w.webp',
      'images/image-828w.png',
      'images/image-828w.webp',
      'images/image-1080w.png',
      'images/image-1080w.webp',
      'images/image-1200w.png',
      'images/image-1200w.webp',
      'images/image-1920w.png',
      'images/image-1920w.webp',
      'images/image-2048w.png',
      'images/image-2048w.webp',
      'images/image-3840w.png',
      'images/image-3840w.webp',
    ].sort(),
  );

  expect(
    assets.find((a) => a.fileName === 'images/image-640w.png')?.source,
  ).toMatchImageSnapshot();
});

test('custom loader options are supported', async () => {
  const { source, assets } = await compile('image.jpg', {
    w: [100, 200],
    format: ['png'],
    name: 'test-[width].[ext]',
    include: '**/*.jpg',
  });

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual([
    'images/test-100.png',
    'images/test-200.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

test('custom query params are supported', async () => {
  const { source, assets } = await compile('image.jpg?w=100;200&format=png', {
    include: /^[^?]+\.jpg(\?.*)?$/,
  });

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual([
    'images/image-100w.png',
    'images/image-200w.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

test('imagetools params are supported', async () => {
  const { source, assets } = await compile(
    'image.jpg?w=100;200&format=png&grayscale',
    {
      include: /^[^?]+\.jpg(\?.*)?$/,
    },
  );

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual([
    'images/image-100w.png',
    'images/image-200w.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

describe('LQIP', function () {
  //   test('color LQIP is supported', async () => {
  //     const { stats } = await compiler(
  //       'fixtures/image.jpg?responsive',
  //       _dirname,
  //       {
  //         lqip: { type: 'color' },
  //       },
  //     );

  //     expect(stats.modules).toBeDefined();
  //     expect(stats.modules![0]?.modules).toHaveLength(3);

  //     const output = stats.modules?.[0]?.modules?.[0]?.source;
  //     expect(sanitizeOutput(output)).toMatchSnapshot();
  //   });

  //   test('inline LQIP is supported', async () => {
  //     const { stats } = await compiler(
  //       'fixtures/image.jpg?responsive',
  //       _dirname,
  //       {
  //         lqip: { type: 'inline' },
  //       },
  //     );

  //     expect(stats.modules).toBeDefined();
  //     expect(stats.modules![0]?.modules).toHaveLength(3);

  //     const output = stats.modules?.[0]?.modules?.[0]?.source;
  //     expect(sanitizeOutput(output)).toMatchSnapshot();
  //   });

  test('blurhash LQIP is supported', async () => {
    const { source } = await compile('image.jpg', {
      include: '**/*.jpg',
      w: [100, 200],
      lqip: { type: 'blurhash' },
    });

    expect(source).toMatchSnapshot();
  });
});
