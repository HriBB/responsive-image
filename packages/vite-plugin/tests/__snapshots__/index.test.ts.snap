// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it produces expected output 1`] = `
"function r(i, a, m) {
  const w = i.map((e) => e.width).reduce((e, g) => g >= a && e >= a ? g >= e ? e : g : g >= e ? g : e, 0);
  return i.find(
    (e) => e.width === w && e.format === m
  );
}
const n = /* @__PURE__ */ new Map();
function p(i) {
  return n.get(i);
}
var t;
(t = p("env")) == null || t.deviceWidths;
const o = [{ url: "/images/image-640w.png", width: 640, format: "png" }, { url: "/images/image-640w.webp", width: 640, format: "webp" }, { url: "/images/image-750w.png", width: 750, format: "png" }, { url: "/images/image-750w.webp", width: 750, format: "webp" }, { url: "/images/image-828w.png", width: 828, format: "png" }, { url: "/images/image-828w.webp", width: 828, format: "webp" }, { url: "/images/image-1080w.png", width: 1080, format: "png" }, { url: "/images/image-1080w.webp", width: 1080, format: "webp" }, { url: "/images/image-1200w.png", width: 1200, format: "png" }, { url: "/images/image-1200w.webp", width: 1200, format: "webp" }, { url: "/images/image-1920w.png", width: 1920, format: "png" }, { url: "/images/image-1920w.webp", width: 1920, format: "webp" }, { url: "/images/image-2048w.png", width: 2048, format: "png" }, { url: "/images/image-2048w.webp", width: 2048, format: "webp" }, { url: "/images/image-3840w.png", width: 3840, format: "png" }, { url: "/images/image-3840w.webp", width: 3840, format: "webp" }], s = {
  imageTypes: ["png", "webp"],
  availableWidths: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
  aspectRatio: 2,
  imageUrlFor(i, a) {
    var m;
    return (m = r(o, i, a ?? "png")) == null ? void 0 : m.url;
  }
};
console.log(s);
"
`;
